contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -11,66 +11,150 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
+  /// @notice Contains the signing address of each oracle
+  struct OracleSet {
+    uint8 f; // Number of faulty nodes allowed
+    address[] signers;
+    mapping(address => uint256) _positions; // 1-indexed to detect unset values
+  }
+  mapping(bytes4 donId => OracleSet) internal s_configs;
 
-  mapping(bytes32 => address) internal s_reports;
+  // reportId = keccak256(bytes20(receiver) | selector | workflowExecutionId)
+  mapping(bytes32 reportId => address transmitter) internal s_reports;
 
   constructor() ConfirmedOwner(msg.sender) {}
 
-  // send a report to targetAddress
+  uint256 internal constant MAX_ORACLES = 31;
+
+  // NOTE: we don't inherit OCR2Base since unlike aggregator we only care about signers, not transmitters
+  // and the signers don't fetch their config from the forwarder
+  function setConfig(bytes4 donId, uint8 f, address[] calldata signers) external nonReentrant {
+    if (f == 0) revert FaultToleranceMustBePositive();
+    if (signers.length > MAX_ORACLES) revert ExcessSigners(signers.length, MAX_ORACLES);
+    if (signers.length <= 3 * f) revert InsufficientSigners(signers.length, 3 * f + 1);
+
+    // TODO: how does setConfig handle expiration? e.g. if the signer set changes
+
+    // remove any old signer addresses
+    for (uint256 i = 0; i < s_configs[donId].signers.length; ++i) {
+      address signer = s_configs[donId].signers[i];
+      delete s_configs[donId]._positions[signer];
+    }
+    s_configs[donId].signers = signers;
+
+    // add new signer addresses
+    for (uint256 i = 0; i < signers.length; ++i) {
+      // assign indices, detect duplicates
+      address signer = signers[i];
+      if(s_configs[donId]._positions[signer] != 0) revert RepeatedSigner();
+      s_configs[donId]._positions[signer] = uint8(i) + 1;
+      s_configs[donId].signers.push(signer);
+    }
+    s_configs[donId].f = f;
+  }
+
+  function getSelector(bytes memory _data) private pure returns(bytes4 sig) {
+      assembly {
+          sig := mload(add(_data, 32))
+      }
+  }
+
+  // send a report to receiver
   function report(
-    address targetAddress,
+    address receiver,
     bytes calldata data,
     bytes[] calldata signatures
   ) external nonReentrant returns (bool) {
-    if (data.length < SELECTOR_LENGTH + REPORT_LENGTH) {
+    if (data.length < SELECTOR_LENGTH + Utils.REPORT_HEADER_LENGTH) {
       revert InvalidData(data);
     }
 
     // data is an encoded call with the selector prefixed: (bytes4 selector, bytes report, ...)
     // we are able to partially decode just the first param, since we don't know the rest
+    bytes4 selector = getSelector(data);
     bytes memory rawReport = abi.decode(data[4:], (bytes));
 
-    // TODO: we probably need some type of f value config?
+    (/* bytes32 workflowId */, bytes4 donId, bytes32 workflowExecutionId) = Utils._splitReport(rawReport);
+
+    if (signatures.length != s_configs[donId].f + 1) {
```

Comment:
```suggestion
    if (signatures.length < s_configs[donId].f + 1) {
```

We could get more than f+1 signatures, no?

Reply:
libocr's version is strict in the number of signatures expected, though mostly so we don't reimburse gas costs for redundant signature verifications